### DISCUSSION TITLE
"refusing to allow a bot to create or update workflow" when pushing tags on older commits

### DISCUSSION THREAD

**Discussion by bewuethr** [community/community#26164] [Actions]
<p>From within an <a href="https://github.com/bewuethr/release-tracker-action" rel="nofollow noopener">action</a>, I’m trying to modify tags to automate the versioning scheme recommended in the <a href="https://github.com/actions/toolkit/blob/master/docs/action-versioning.md" rel="nofollow noopener">toolkit documentation</a>: a “most recent major release” tag (such as v1) pointing at the most recent semantic versioning tag with the same major release, i.e., if I have tags v1.0.0, v1.0.1, v1.1.0, v1.2.0 and v1.2.1, then v1 should point to the same commit as v1.2.1. When I release v1.3.0, v1 should automatically be updated to point to what v1.3.0 points to.</p>
<p>To have access to all the tags, I check out the repository with all commits:</p>
<pre><code>- name: 'Check out code'
      uses: 'actions/checkout@v2'
      with:
        fetch-depth: 0
</code></pre>
<p>The action runs a shell script in a Docker container, and I set the GITHUB_TOKEN in the environment:</p>
<pre><code>- name: 'Update release tags for latest major and minor releases'
      uses: 'bewuethr/release-tracker-action@master'
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
</code></pre>
<p>According to the <a href="https://github.com/actions/checkout#checkout-v2" rel="nofollow noopener">documentation for the checkout action</a>, the auth token is persisted in the Git config, and indeed, I can do things like</p>
<pre><code>git push --delete origin "$tag"
</code></pre>
<p>without problems.</p>
<p>However, after creating new tags, if any new tag doesn’t point to the most recent commit, I can’t push it and get an error message like this:</p>
<pre><code>! [remote rejected] v0.1 -&gt; v0.1 (refusing to allow a bot to create or update workflow `.github/workflows/releasetracker.yml`)
</code></pre>
<p>I’m trying to push nothing but the tags by using something like</p>
<pre><code>git push origin "$tag1" "$tag2"
</code></pre>
<p>and definitely didn’t modify the workflow file.</p>
<p><strong>From a shell script running in a docker action, how can I push my new tags pointing to older commits?</strong></p>
<p>Things I’ve tried:</p>
<ul>
<li>Use the authentication token explicitly in the URL for the remote by setting it to <a class="inline-onebox-loading">https://$GITHUB_ACTOR:$GITHUB_TOKEN@github.com/$GITHUB_REPOSITORY</a> or just <a class="inline-onebox-loading">https://$GITHUB_TOKEN@github.com/$GITHUB_REPOSITORY</a>
</li>
<li>Use SSH instead of HTTPS by setting the remote URL to <a href="mailto:git@github.com">git@github.com</a>:$GITHUB_REPOSITORY</li>
<li>Add .git to remote URL</li>
<li>Push directly from a step in the job instead of from within the action with run: git push --tags</li>
<li>Use branches instead of moving tags (see also <a href="https://github.com/actions/toolkit/issues/214" rel="nofollow noopener">this discussion</a>)</li>
</ul>
<p>I’ve seen other people doing similar things by using the GitHub API instead of Git commands (like in the <a href="https://github.com/jessfraz/branch-cleanup-action" rel="nofollow noopener">Branch Cleanup Action</a>, where a branch is deleted using an <a href="https://github.com/jessfraz/branch-cleanup-action/blob/master/cleanup-pr-branch#L80-L85" rel="nofollow noopener">API call</a>), but is that the only way? It does feel like a bug that my push is prevented with a reason that doesn’t seem to apply to my situation.</p>
<p>I also haven’t tried using a personal access token instead of GITHUB_TOKEN, because again, it feels like it shouldn’t be necessary.</p>
<hr>
<p>As a side note, posts with similar problems in this forum such as <a href="https://github.community/t/refusing-to-allow-an-integration-to-create-or-update/16326">Refusing to allow an integration to create or update</a> seemed to have “refresh the token of your client” as the solution, but that doesn’t apply in my case.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/bewuethr">@bewuethr</a> ,</p>
<p>The error is due to that GITHUB_TOKEN is used to update the tag. Please create a personal access token as secrets instead. Please follow below items to fix your code:</p>
<ol>
<li>In the workflow yaml, add ‘persist-credentials: false’ for ‘actions/checkout@v2’.</li>
</ol>
<p>Use PAT for ‘bewuethr/release-tracker-action@master’.</p>
<p>Code sample as below:</p>
<pre><code>jobs:
  release-tracking:
    runs-on: 'ubuntu-latest'
    steps:
    - name: 'Check out code'
      uses: 'actions/checkout@v2'
      with:
        fetch-depth: 0
        persist-credentials: false
    - name: 'Update release tags for latest major and minor releases'
      uses: '{org}/release-tracker-action@master'
      env:
        GITHUB_ACTOR: username
        TOKEN: ${{ secrets.PAT }}
</code></pre>
<ol start="2">
<li>
<p>In the ‘<a href="https://github.com/bewuethr/release-tracker-action/blob/master/tagupdater" rel="nofollow noopener">tagupdater</a>’ file, set remote url.</p>
<p>git remote set-url --push origin “<a class="inline-onebox-loading">https://$GITHUB_ACTOR:$TOKEN@github.com/$GITHUB_REPOSITORY</a>”</p>
</li>
</ol>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/6e297b331c348a1734c4d04336d12198a5fe0f89.png" data-download-href="/uploads/short-url/fIxou3enHW6fh5Fl01EPF7Q33Pr.png?dl=1" title="tagpush.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181095890-63a7e13f-feee-4daa-8079-a90831c9d088.png" alt="tagpush.png" data-base62-sha1="fIxou3enHW6fh5Fl01EPF7Q33Pr" width="690" height="249" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/6e297b331c348a1734c4d04336d12198a5fe0f89_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">tagpush.png</span><span class="informations">817×296 17.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>But there is a limitation that a new workflow is triggered again since PAT update the tag, and the workflow will fail.</p>
<p>Probably api call can work better.</p>
<p>Thanks.</p>

**Comment by bewuethr** [comment]
<p>Thanks, I half suspected I’d have to do it like that. I was mostly confused because <em>some</em> things such as deleting tags worked, and the error message “refusing to allow a bot to create or update workflow” is implying that I’m trying to do something with my worfklow, but I wasn’t.</p>

**Comment by antonosmond** [comment]
I hit the same problem. I'm trying to force push tags from a github action and get the same error about updating the workflow which I'm definitely not trying to do.
@bewuethr don't suppose you found a nicer solution than the one recommended here did you?

**Reply by bewuethr** [reply]
I have since switched the action to a composite step action instead of using docker, but more importantly, I use the GitHub API to create the tag instead of pushing it using `git`:

```sh
# Create an annotated tag
createannotatedtag() {
	local tag=$1
	local semvertag=$2


	# Find SHA of object tag points to
	local sha
	sha=$(getparentsha "$semvertag")


	# Create tag object
	echo "Creating tag object for $tag..."
	local tagsha
	tagsha=$(gh api 'repos/{owner}/{repo}/git/tags' \
		--raw-field "tag=$tag" \
		--raw-field "message=latest release in $tag" \
		--raw-field "object=$sha" \
		--raw-field "type=commit" \
		--jq '.sha')


	# Create tag reference
	echo "Creating tag reference..."
	gh api 'repos/{owner}/{repo}/git/refs' \
		--raw-field "ref=refs/tags/$tag" \
		--raw-field "sha=$tagsha" \
		| jq .
}
```

See the code [here](https://github.com/bewuethr/release-tracker-action/blob/ad642f1fe09c2d29f1e68d6a59147d01a275281a/tagupdater#L20-L45). This works without requiring a personal access token; if you default your actions to read only, you have to explicitly add `contents: write` permissions to the workflow (see [docs](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs)).
